Rewrite documentation. (#23305)
jhipster/generator-jhipster: ARCHITECTURE.md

# JHipster architecture

## Yeoman basis

> Internally JHipster uses [yeoman](https://yeoman.io) as core.

Reviewer: ```suggestion
Internally, JHipster uses [Yeoman](https://yeoman.io) as the core. JHipster is the [most popular generator of all time](https://yeoman.io/generators/). 
```

> ## File structure

Reviewer 2: Maybe we can add some kind of glossary, what is a generator, what is the cli, what is jdl? Guess that might help to understand all the different constructs.

- `.blueprint` - development blueprint, used to generate and manage samples
- `.devcontainer` - vscode's devcontainer definitions
- `.github` - github configuration
- `.vscode` - vscode's configuration
- `bin` - jit executable and helper
- `cli` - (exported) cli implementation
- `generators/*` - (exported) generators
  - `command.m[tj]s` - cli options, cli arguments definitions
  - `generator.m[tj]s` - generator implementation
  - `index.m[tj]s` - generator exports. Must re-export generator as default export and the command
  - `internal` - non-exported supporting libs
  - `resources` - supporting resources
  - `jdl` - generator's jdl specifications
  - `support` - (exported) exported supporting libs
  - `templates` - templates folder
- `jdl` - (exported) jdl parser implementation
- `rfcs` - (RFCs)[CONTRIBUTING.md#rfcs]
- `test` - package tests
- `test-integration` - CI related stuff. Samples, scripts.

## Lifecycle

> - [Cli entrypoint](https://github.com/jhipster/generator-jhipster/blob/main/cli/jhipster.cjs)

Reviewer: ```suggestion
- [CLI entry point](https://github.com/jhipster/generator-jhipster/blob/main/cli/jhipster.cjs)
```

- [Basic environment validation](https://github.com/jhipster/generator-jhipster/blob/main/cli/cli.mjs)

> - [Cli arguments parsing and Environment bootstrap](https://github.com/jhipster/generator-jhipster/blob/main/cli/program.mjs)

Reviewer: ```suggestion
- [CLI arguments parsing and environment bootstrapping](https://github.com/jhipster/generator-jhipster/blob/main/cli/program.mjs)
```

>   - Lookup for Generators and Blueprints

Reviewer: ```suggestion
  - Lookup for generators and blueprints
```

>   - Build cli options and arguments definition

Reviewer: ```suggestion
  - Build CLI options and arguments definition
```

  - Parse options and arguments

> - Run Generator (start the Environment) passing options and arguments

Reviewer: ```suggestion
- Run generator (start the environment) passing options and arguments
```

- Run every task from the highest precedence priority until there is no more pending task

### Priorities

#### Initializing (yeoman)

Initial generator information.

- say generator specific hello
- initial environment checks
- argument and options loading

```
  get [Generator.INITIALIZING]() {
    return this.asInitializingTaskGroup() {
      sayHelloTask() {
        this.log.info('Welcome to your generator');
      },
      envChecks() {
        checkNode();
        checkDocker();
      },
      async loadOptions() {
        this.parseJHipsterArguments(command.arguments);
        this.parseJHipsterOptions(command.options);
      },
    }
  }

> ```

Reviewer: Outdent, so it's easier to copy/paste.

#### Prompting (yeoman)

Prompt for configuration.

TODO

#### Configuring (yeoman)

> Check and fix configurations.

Reviewer: ```suggestion
Check and fix configurations:
```

```
  get [Generator.CONFIGURING]() {
    return this.asConfiguringTaskGroup() {
      checkConfig() {
        if (this.jhipsterConfigWithDefaults.reactive && this.jhipsterConfigWithDefaults.cacheProvider !== 'no') {
          this.log.warn("Reactive applications doesn't support cache. Disabling");
          this.jhipsterConfig.cacheProvider = 'no';
        }
      },
    }
  }

> ```

Reviewer: I think it looks better when these code blocks are outdented and left-aligned. Also, add `js` as the language.

#### Composing (base)

> Compose with other generators.

Reviewer: ```suggestion
Compose with other generators:
```

```
  get [Generator.COMPOSING]() {
    return this.asComposingTaskGroup() {
      composing() {
        if (this.jhipsterConfigWithDefaults.clientFramework === 'angular') {
          this.composeWithJHipster('angular');
        }
      },
    }
  }
```

#### Loading (base)

> Load configuration.

Reviewer: ```suggestion
Load configuration:
```

```
  get [Generator.LOADING]() {
    return this.asLoadingTaskGroup() {
      loading({ application }) {
        application.myCustomConfig = this.jhipsterConfig.myCustomConfig;
        application.myBlueprintCustomConfig = this.blueprintConfig.myBlueprintCustomConfig;
      },
    }
  }
```

#### Preparing (base)

> Generate properties to improve understanding.

Reviewer: ```suggestion
Generate properties to improve understanding:
```

```
  get [Generator.PREPARING]() {
    return this.asPreparingTaskGroup() {
      preparing({ application }) {
        application.myCustomConfigFoo = this.jhipsterConfig.myCustomConfig === 'foo';
        application.myCustomConfigBar = this.jhipsterConfig.myCustomConfig === 'bar';
        application.myCustomConfigNo = !this.jhipsterConfig.myCustomConfig || this.jhipsterConfig.myCustomConfig === 'no';
      },
    }
  }
```

#### Configuring each entity (base-application)

> Configure and check entity's configuration.

Reviewer: ```suggestion
Configure and check entity's configuration:
```

```
  get [Generator.CONFIGURING_EACH_ENTITY]() {
    return this.asConfiguringEachEntityTaskGroup() {
      configuring({ application, entityConfig }) {
        if (application.searchEngineNo && entityConfig.searchEngine && entityConfig.searchEngine !== 'no') {
          this.log.warn("Search engine cannot be enabled at entity because it's disabled at application");
          entityConfig.searchEngine = 'no';
        }
      },
    }
  }
```

#### Loading entities (base-application)

Usually empty the entire entity configuration is loaded by default.

#### Preparing each entity (base-application)

> Generate properties to improve understanding at entity level.

Reviewer: ```suggestion
Generate properties to improve understanding at entity level:
```

```
  get [Generator.PREPARING_EACH_ENTITY]() {
    return this.asPreparingEachEntityTaskGroup() {
      preparing({ application, entity }) {
        entity.dtoMapstruct = entity.dto === 'mapstruct';
      },
    }
  }
```

#### Preparing each entity field (base-application)

> Generate properties to improve understanding at field level.

Reviewer: ```suggestion
Generate properties to improve understanding at the field level:
```

```
  get [Generator.PREPARING_EACH_ENTITY_FIELD]() {
    return this.asPreparingEachEntityFieldTaskGroup() {
      preparing({ application, entity, field }) {
        field.technologyFieldTypeIntegerMap = field.fieldType === 'Integer';
      },
    }
  }
```

#### Preparing each entity relationship (base-application)

> Generate properties to improve understanding at relationship level.

Reviewer: ```suggestion
Generate properties to improve understanding at the relationship level:
```

```
  get [Generator.PREPARING_EACH_ENTITY_RELATIONSHIP]() {
    return this.asPreparingEachEntityRelationshipTaskGroup() {
      preparing({ application, entity, relationship }) {
        relationship.technologyRelationshipDbName = relationship.relationshipTypeOneToOne ? 'foo' : 'bar';
      },
    };
  }
```

#### Default (yeoman)

> Generate properties to improve understanding that depends on others properties.

Reviewer: ```suggestion
Generate properties to improve understanding that depends on others' properties:
```

```
  get [Generator.DEFAULT]() {
    return this.asDefaultTaskGroup() {
      preparing({ application, entities }) {
        application.hasEntityFieldInteger = entities.some(entity => entity.fields.some(field => field.fieldTypeInteger));
      },
    };
  }
```

#### Writing (yeoman)

> Write files to in memory file system.

Reviewer: ```suggestion
Write files to in-memory file system.
```

> There are a lot of apis to write files, copy files, delete files.

Reviewer: ```suggestion
There are a lot of APIs to write files, copy files, and delete files.
```

> `writeFiles` is the most used at official generators.

Reviewer: ```suggestion
The `writeFiles()` method is the most used in official generators.
```

```
  get [Generator.WRITING]() {
    return this.asWritingTaskGroup({
      writingTask({ application }) {
        this.writeFiles({
          blocks: [
            {
              condition: ctx => ctx.shouldWrite,
              templates: ['template.file'],
            }
          ],
          context: application,
        });
      },
    });
  }
```

#### Writing entities (base-application)

> Write entity files to in memory file system.

Reviewer: ```suggestion
Write entity files to the in-memory file system.
```

> Entities writing is a separed priority to keep workflow sane when using options like `--skip-application` and `--single-entity`.

Reviewer: ```suggestion
Writing entities is a separate priority to keep the workflow sane when using options like `--skip-application` and `--single-entity`.
```

```
  get [Generator.WRITING_ENTITIES]() {
    return this.asWritingTaskGroup({
      writingTask({ application, entities }) {
        for (const entity of entities) {
          this.writeFiles({
            blocks: [
              {
                condition: ctx => ctx.shouldWrite,
                templates: ['entity.template.file'],
              }
            ],
            context: { ...application, ...entity },
          });
        }
      },
    });
  }
```

#### Post writing (base)

> Injects code in the generated source

Reviewer: ```suggestion
Injects code in the generated source.
```

##### Injecting code with provided apis (needles)

> JHipster adds apis to some code injection.

Reviewer: ```suggestion
JHipster adds APIs to some code injection:
```

Should this be "to do some code injection"?

```
  get [Generator.POST_WRITING]() {
    return this.asPostWritingTaskGroup({
      postWritingTask({ source }) {
        source.someProvidedInjectionApi({ code: 'some code' });
      }
    });
  }
```

##### Custom code injection

Every file can be edited manually.

```
  get [Generator.POST_WRITING]() {
    return this.asPostWritingTaskGroup({
      postWritingTask({ source }) {
        this.editFile('path/to/some/file', content => content.replaceAll('some content', 'another content'));
      }
    });
  }
```

#### Install (yeoman)

Usually empty.
Install task is queued by detecting `package.json` changes.

#### End (yeoman)

> Print generator result and info.

Reviewer: ```suggestion
Print generator result and info:
```

# JHipster architecture

## Yeoman basis

Internally JHipster uses [yeoman](https://yeoman.io) as core.

## File structure

- `.blueprint`          - development blueprint, used to generate and manage samples 
- `.devcontainer`       - vscode's devcontainer definitions
- `.github`             - github configuration
- `.vscode`             - vscode's configuration
- `bin`                 - jit executable and helper
- `cli`                 - (exported) cli implementation
- `generators/*`        - (exported) generators
  - `generator.m[tj]s`  - generator implementation
  - `index.m[tj]s`      - generator exports. Must re-export generator as default export.
  - `internal`          - non-exported supporting libs
  - `resources`         - supporting resources
  - `jdl`               - generator's jdl specifications
  - `support`           - (exported) exported supporting libs
  - `templates`         - templates folder
- `jdl`                 - (exported) jdl parser implementation

> - `rfcs`                - (RFCs)[CONTRIBUTING.md#rfcs]

Reviewer 3: the rfc thing never really took off, so may be we should just remove it
